File: 脑电特征py/tiqutz12.py
import numpy as np
from scipy import signal
from scipy.signal import spectrogram, hamming
# 假设以下函数已定义（基于之前的转换）
from center import center  # center.m 转换的Python函数
from FAA import FAA  # FAA.m 转换的Python函数
from LZC import LZC  # LZC.m 转换的Python函数
from SampEn import SampEn  # SampEn.m 转换的Python函数

def tiqutz12(EEG):
    """
    从EEG数据中提取特征集。

    参数:
    EEG : numpy.ndarray
        输入的EEG数据矩阵，行表示时间点，列表示通道。

    返回:
    featureSet : numpy.ndarray
        提取的特征集，形状为 (通道数, 特征维度)。
    """
    feat_dim = 7  # 特征维度
    if EEG.size != 0:  # 检查EEG是否为空
        len_, wei = EEG.shape
        if len_ < wei:  # 如果行数小于列数，矩阵转置
            EEG = EEG.T
            len_, wei = EEG.shape  # 更新维度
    else:
        len_, wei = 0, 0

    featureSet = np.zeros((wei, feat_dim))  # 初始化特征集
    featureSet_tmp = np.zeros((wei, feat_dim))  # 临时特征集

    # 根据数据长度初始化零填充向量
    if len_ == 500:
        z = np.zeros((12, 1))
    else:
        z = np.zeros((36, 1))

    len_size = len_ / 250  # 计算时间窗口大小

    for chn in range(wei):  # 遍历每个通道
        user_fp = np.vstack([EEG[:, chn].reshape(-1, 1), z])[:, 0]  # 信号末尾补零

        # 调用特征提取函数（假设已定义）
        c01, c0_average = c0complex(EEG[:, chn], 250, 0, len_size)
        D_inf_all, D_q_0all, D_q_1all1, average_D_inf, average_D_q_0, average_D_q_1 = Renyi_spectral(EEG[:, chn].T, 250, 0, len_size)
        Pxx1, F1 = fftpsd1(user_fp, 250)

        # 计算频谱相关特征
        ct1 = center(Pxx1, F1)  # 频谱中心
        max1 = np.max(Pxx1)  # 最大功率
        mean1 = np.sum(Pxx1) / 129  # 平均功率

        # 提取alpha波段（假设alpha函数已定义）
        EEG_band = alpha(EEG[:, chn])

        # 计算FAA、LZC和样本熵
        faa = FAA(EEG)
        lzc = LZC(EEG[:, chn])
        r = 0.2 * np.std(EEG[:, chn])  # 容差
        SEn = SampEn(2, r, EEG[:, chn], 1)  # 样本熵

        # 组合特征
        feat = [faa, ct1, max1, mean1, lzc, SEn, D_q_1all1]
        featureSet_tmp[chn, :] = feat  # 存储到临时特征集

    featureSet = featureSet_tmp  # 更新特征集
    return featureSet


def EEGRhythm_alpha(x):
    """
    提取EEG信号的alpha节律。

    参数:
    x : numpy.ndarray
        输入的单通道EEG信号。

    返回:
    eeg_rhy_alpha : list
        包含不同频段节律的列表，其中第三个元素为alpha节律。
    """
    # 注意：这里需要IterFiltMulti和Settings_IF_v1的Python实现
    # 以下为简化实现，假设使用外部库（如mne）提取alpha节律
    x = x.T  # 转置信号
    Fs = 250  # 采样率

    # 模拟IterFiltMulti，实际需替换为具体实现
    opt = {'IF.Xi': 1.6, 'IF.alpha': 'ave', 'IF.delta': 0.001, 'IF.NIMFs': 20}
    MIMF = IterFiltMulti(x, opt)  # 假设IterFiltMulti已定义
    eeg_rhy_alpha = EEGRhythm(MIMF, Fs)
    return eeg_rhy_alpha


def EEGRhythm(mIMF, Fs):
    """
    根据MIMF分解结果提取EEG的频段节律。

    参数:
    mIMF : list
        迭代滤波分解后的MIMF结果。
    Fs : float
        采样频率。

    返回:
    eeg_rhy : list
        包含delta、theta、alpha、beta、gamma节律的列表。
    """
    N = mIMF[0][1].shape[1]  # 数据长度
    N_ch = mIMF[0][0].shape[0]  # 通道数
    N_mimf = len(mIMF[0])  # MIMF数量

    # 计算每个MIMF的平均频率
    mf = np.zeros(N_mimf)
    for imf_ind in range(N_mimf):
        # 使用scipy.signal的频谱分析计算平均频率
        freqs, psd = signal.welch(mIMF[0][imf_ind].T, fs=Fs, nperseg=min(256, N))
        mf[imf_ind] = np.sum(freqs * psd) / np.sum(psd) if np.sum(psd) != 0 else 0
    mf = np.mean(mf, axis=0)

    # 分配频段
    delta_i, theta_i, alpha_i, beta_i, gamma_i = [], [], [], [], []
    for i in range(len(mf)):
        if 0.1 < mf[i] < 4:
            delta_i.append(i)
        elif 4 <= mf[i] < 8:
            theta_i.append(i)
        elif 8 <= mf[i] < 14:
            alpha_i.append(i)
        elif 14 <= mf[i] < 30:
            beta_i.append(i)
        elif 30 <= mf[i] < 95:
            gamma_i.append(i)

    # 提取各频段节律
    eeg_rhy = [
        EEG_Rhy(mIMF, delta_i),
        EEG_Rhy(mIMF, theta_i),
        EEG_Rhy(mIMF, alpha_i),
        EEG_Rhy(mIMF, beta_i),
        EEG_Rhy(mIMF, gamma_i)
    ]
    return eeg_rhy


def EEG_Rhy(mIMF, ind):
    """
    根据指定索引提取特定频段的节律。

    参数:
    mIMF : list
        MIMF分解结果。
    ind : list
        需要提取的MIMF索引。

    返回:
    Rhy : numpy.ndarray
        指定频段的节律信号。
    """
    Rhy = np.zeros_like(mIMF[0][0])
    for i in ind:
        Rhy += mIMF[0][i]
    return Rhy


# 假设未定义的函数（需自行实现或替换）
def c0complex(data, fs, flag, len_size):
    """
    占位函数：计算复杂性特征（需自行实现）。
    """
    return 0, 0


def Renyi_spectral(data, fs, flag, len_size):
    """
    占位函数：计算Renyi谱特征（需自行实现）。
    """
    return 0, 0, 0, 0, 0, 0


def fftpsd1(data, fs):
    """
    占位函数：计算功率谱密度（示例实现）。
    """
    freqs, psd = signal.welch(data, fs=fs, nperseg=min(256, len(data)))
    return psd, freqs


def alpha(data):
    """
    占位函数：提取alpha波段（示例实现）。
    """
    freqs, psd = signal.welch(data, fs=250, nperseg=min(256, len(data)))
    alpha_band = psd[(freqs >= 8) & (freqs < 14)]
    return alpha_band


def IterFiltMulti(x, opt):
    """
    占位函数：迭代滤波分解（需自行实现）。
    """
    # 示例：返回简单的分解结果
    return [[x, x]]  # 需替换为实际实现

File: 脑电特征mat/tiqutz12.m
function [featureSet]=tiqutz12(EEG)
%printf('128');
feat_dim=7;
%EEG=control{1}{1};
if ~isempty(EEG)
[len,wei]=size(EEG);
if len<wei
   EEG=EEG';
  % wei=len;
end
end
[len,wei]=size(EEG);
featureSet=zeros(1,feat_dim);
featureSet_tmp=zeros(wei,feat_dim);
 if len==500
    z=zeros(12,1);
    else
        z=zeros(36,1);
 end
 len_size=len/250;

 
for chn=1:wei
   
    user_fp=[EEG(:,chn);z];
    [c01,c0_average]=c0complex(EEG(:,chn),250,0, len_size);
    %[Correlation1,M_C]=reC(EEG(:,chn),250,0, len_size);
    [D_inf_all,D_q_0all,D_q_1all1,average_D_inf,average_D_q_0,average_D_q_1]=Renyi_spectral(EEG(:,chn)',250,0, len_size);
    [Pxx1,F1] = fftpsd1(user_fp,250);

    ct1=center(Pxx1,F1);
    max1=max(Pxx1);
    mean1=sum(Pxx1)/129;
    EEG_band=alpha(EEG(:,chn));
%     EEG_band = EEGRhythm_alpha(EEG(:,chn));
%     apv=APV(EEG_band(400:2000,:),EEG_Orig(:,chn));
    faa = FAA(EEG);
    lzc=LZC(EEG(:,chn));
    r=0.2*std(EEG(:,chn));
    SEn = SampEn(2,r,EEG(:,chn),1);
%     feat=[c01,D_q_1all1,ct1,max1,mean1,faa,lzc,SEn];%[c01,Correlation1,D_q_1all1,ct1,max1,mean1];
    feat=[faa,ct1,max1,mean1,lzc,SEn,D_q_1all1];%[c01,Correlation1,D_q_1all1,ct1,max1,mean1];
    featureSet1(1,(chn-1)*feat_dim+1:chn*feat_dim)=feat;
    featureSet_tmp(chn,:)=feat;
%     featureSet=featureSet+feat;
    featureSet=featureSet_tmp;
end
end
%%%%%%%%%%%%-----------计算 频段  alpha beta

function [eeg_rhy_alpha]= EEGRhythm_alpha(x)
addpath(genpath('../Preprocessing/MIF-main/'));
% addpath(genpath('Acupuncture/'))
% EEG_rhy=cell(1,73);
% for i =1 : 73
    


% x = EEG1{1,i}'; % Take any multi-channel EEG signal in x
x = x';
Fs = 250;
opt=Settings_IF_v1('IF.Xi',1.6,'IF.alpha','ave','IF.delta',.001,'IF.NIMFs',20);
MIMF = IterFiltMulti(x,opt);
eeg_rhy_alpha = EEGRhythm(MIMF,Fs);
% EEG_rhy{1,i} = eeg_rhy;
% end
end


function [eeg_rhy]= EEGRhythm(mIMF,Fs)


N=size(mIMF{1,2},2); % Length of the data
N_ch=size(mIMF{1,1},1); % number of channel
N_mimf = size(mIMF,2); % Number of MIMF

for imf_ind = 1: N_mimf
    mf(:,imf_ind)=meanfreq(mIMF{1,imf_ind}', Fs);
end
mf = mean(mf,1);

delta_i=[];theta_i=[];alpha_i=[];
beta_i=[];gamma_i=[];
for i=1:length(mf) % Select IMF belongs to band
    if ((mf(i)>.1) && (mf(i) <4))
        delta_i=[delta_i,i];
    elseif (mf(i)>=4 && mf(i) <8)
        theta_i=[theta_i,i];
    elseif (mf(i)>=8 && mf(i) <14)
        alpha_i=[alpha_i,i];
    elseif (mf(i)>=14 && mf(i) <30)
        beta_i=[theta_i,i];
    elseif (mf(i)>=30 && mf(i) <95)
        gamma_i=[gamma_i,i];
    end
end

eeg_rhy{1,1} = EEG_Rhy(mIMF, delta_i);
eeg_rhy{1,2} = EEG_Rhy(mIMF, theta_i);
eeg_rhy{1,3} = EEG_Rhy(mIMF, alpha_i);
eeg_rhy{1,4} = EEG_Rhy(mIMF, beta_i);
eeg_rhy{1,5} = EEG_Rhy(mIMF, gamma_i);


end


% Axulary functions

function [Rhy] = EEG_Rhy(mIMF, ind)
Rhy = zeros(size(mIMF{1,1}));
for i = 1: length(ind)
    Rhy = Rhy + mIMF{1,ind(i)};
end
end



%featureSet=featureSet+feat;